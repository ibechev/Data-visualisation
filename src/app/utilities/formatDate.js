export const formatDate = date => {
  const months = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec"
  ];
  const tempDate = date.split("-");
  return `${months[parseInt(tempDate[1]) - 1]} ${tempDate[0].substring(2)}`;
};
