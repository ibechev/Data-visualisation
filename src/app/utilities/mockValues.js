export const mockAlerts = [
  {
    name: "Adam Campton",
    event: "end of year review",
    eventDate: "2019-01-18",
    importance: "very"
  },
  {
    name: "Emma O'Sullivan",
    event: "end of year review",
    eventDate: "2019-01-03",
    importance: "low"
  },
  {
    name: "Brandon Walsh",
    event: "end of year review",
    eventDate: "2019-02-06",
    importance: "middle"
  },
  {
    name: "David O'Brien",
    event: "end of year review",
    eventDate: "2019-02-23",
    importance: "very"
  },
  {
    name: "Sophie O'Connor",
    event: "end of year review",
    eventDate: "2019-01-03",
    importance: "low"
  },
  {
    name: "Sean Connor",
    event: "end of year review",
    eventDate: "2019-01-05",
    importance: "very"
  },
  {
    name: "Gery Brown",
    event: "end of year review",
    eventDate: "2019-01-12",
    importance: "middle"
  },
  {
    name: "Stephen Greenday",
    event: "end of year review",
    eventDate: "2019-02-11",
    importance: "very"
  },
  {
    name: "Lisa Aniston",
    event: "end of year review",
    eventDate: "2018-12-27",
    importance: "low"
  },
  {
    name: "Gordon Niels",
    event: "end of year review",
    eventDate: "2018-12-28",
    importance: "middle"
  }
];

export const mockCustomerInfo = {
  max: 270,
  data: [
    { month: "Dec-2017", value: 99 },
    { month: "Jan-2018", value: 80 },
    { month: "Feb-2018", value: 71 },
    { month: "Mar-2018", value: 96 },
    { month: "Apr-2018", value: 80 },
    { month: "May-2018", value: 115 },
    { month: "Jun-2018", value: 146 },
    { month: "Jul-2018", value: 121 },
    { month: "Aug-2018", value: 160 },
    { month: "Sep-2018", value: 168 },
    { month: "Oct-2018", value: 185 },
    { month: "Nov-2018", value: 203 }
  ]
};
