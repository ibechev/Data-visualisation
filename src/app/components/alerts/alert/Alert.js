import React from "react";
import PropTypes from "prop-types";
import { formatDate } from "../../../utilities/formatDate";

import LevelBar from "../../levelBar/LevelBar";

const Alert = ({ name, event, eventDate, importance }) => {
  return (
    <li className="alert">
      <div className="info-wrapper">
        <div className="section-primary">
          <span className="icon">
            <i className="fas fa-birthday-cake" />
          </span>
          <p className="name">{name ? name : ""}</p>
        </div>

        <p className="event">
          {name ? `${name.split(" ")[0]}'s ${event} coming up` : ""}
        </p>
      </div>

      <div className="section-stats">
        <p className="date">{eventDate ? formatDate(eventDate) : ""}</p>
        {importance && <LevelBar level={importance} />}
      </div>
    </li>
  );
};

Alert.propTypes = {
  name: PropTypes.string.isRequired,
  event: PropTypes.string.isRequired,
  eventDate: PropTypes.string.isRequired,
  importance: PropTypes.oneOf(["very", "low", "middle"]).isRequired
};

export default Alert;
