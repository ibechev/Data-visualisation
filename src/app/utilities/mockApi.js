import { mockAlerts, mockCustomerInfo } from "../utilities/mockValues";

const randomTimeRange = (min, max) => {
  return (Math.floor(Math.random() * (max - min + 1)) + min) * 1000;
};
const simulateLatency = ms => new Promise(resolve => setTimeout(resolve, ms));

export const mockFetchAlerts = async () => {
  await simulateLatency(randomTimeRange(0, 3));
  return mockAlerts.map(alert => ({ ...alert }));
};

export const mockFetchClientsInfo = async () => {
  const { max, data } = mockCustomerInfo;
  await simulateLatency(randomTimeRange(0, 3));
  return {
    max,
    data: data.map(each => ({ ...each }))
  };
};
