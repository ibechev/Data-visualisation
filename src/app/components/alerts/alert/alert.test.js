import React from "react";
import { mount } from "enzyme";
import Alert from "./Alert";
import renderer from "react-test-renderer";
import { propTypeErrors } from "../../../../../tests/config/testUtils";
import { formatDate } from "../../../utilities/formatDate";

describe("Expect <Alert />", () => {
  const VERY = "very";
  const MIDDLE = "middle";
  const LOW = "low";

  const defaultProps = {
    name: "John Doe",
    event: "Test event",
    eventDate: "Jan-2019",
    importance: VERY
  };

  const setup = (testProps = {}) => {
    const props = {
      ...defaultProps,
      ...testProps
    };
    return mount(<Alert {...props} />);
  };

  it("to render correctly", () => {
    const wrapper = renderer.create(<Alert {...defaultProps} />).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("to not throw error with correct props", () => {
    propTypeErrors(<Alert {...defaultProps} />, defaultProps);
    propTypeErrors(<Alert {...defaultProps} importance={MIDDLE} />, {
      ...defaultProps,
      importance: MIDDLE
    });
    propTypeErrors(<Alert {...defaultProps} importance={LOW} />, {
      ...defaultProps,
      importance: LOW
    });
  });

  describe("when props are correct", () => {
    const wrapper = setup();

    it("to display correct text in <p class='name' />", () => {
      expect(wrapper.find("p.name").text()).toEqual(defaultProps.name);
    });

    it("to display correct text in <p class='event' />", () => {
      expect(wrapper.find("p.event").text()).toEqual(
        `${defaultProps.name.split(" ")[0]}'s ${defaultProps.event} coming up`
      );
    });

    it("to display correct text in <p class='date' />", () => {
      expect(wrapper.find("p.date").text()).toEqual(
        formatDate(defaultProps.eventDate)
      );
    });

    it("to display <LevelBar /> in <div class='section-stats' />", () => {
      expect(wrapper.find("div.section-stats").find("LevelBar")).toHaveLength(
        1
      );
    });
  });
  describe("when props are missing", () => {
    const testProps = {
      name: "",
      event: "",
      eventDate: "",
      importance: VERY
    };

    const wrapper = setup(testProps);

    it("to display  text in <p class='name' />", () => {
      expect(wrapper.find("p.name").text()).toEqual("");
    });

    it("to not display text in <p class='event' />", () => {
      expect(wrapper.find("p.event").text()).toEqual("");
    });

    it("to not display text in <p class='date' />", () => {
      expect(wrapper.find("p.date").text()).toEqual("");
    });
  });
});
