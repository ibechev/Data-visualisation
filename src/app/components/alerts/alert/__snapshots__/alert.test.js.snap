// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <Alert /> to render correctly 1`] = `
<li
  className="alert"
>
  <div
    className="info-wrapper"
  >
    <div
      className="section-primary"
    >
      <span
        className="icon"
      >
        <i
          className="fas fa-birthday-cake"
        />
      </span>
      <p
        className="name"
      >
        John Doe
      </p>
    </div>
    <p
      className="event"
    >
      John's Test event coming up
    </p>
  </div>
  <div
    className="section-stats"
  >
    <p
      className="date"
    >
      undefined n
    </p>
    <div
      className="level-bar-wrapper"
    >
      <p
        className="title"
      >
        very
      </p>
      <div
        className="bars"
      >
        <span
          className="bar low active"
        />
        <span
          className="bar middle active"
        />
        <span
          className="bar very active"
        />
      </div>
    </div>
  </div>
</li>
`;
